### PR TITLE
8264593: debug.cpp utilities should be available in product builds.

### PR DESCRIPTION
**Pull request by kevinjwalls**
This is a pull request to move the ifndef PRODUCT in src/hotspot/share/utilities/debug.cpp, and add JNIEXPORT for symbol visibility on Windows.

With these changes, we can use gdb on a live Linux process of a product build JVM, and call debug.cpp utilities.
e.g. "call universe()", or "call hsfind(0xADDRESS)".

(I found that making calls in the JLI_Launch thread (gdb's thread 1) will crash, but switching first to a JavaThread works.)

In Visual Studio use the Immediate window, just type the function name.  Verified that without the changes, the symbols can't be seen/called.

Printing happens on the LIVE jvm's tty (not the gdb session).

WizardMode is a develop flag, can't be changed in a product build, so avoid changing that in the debug() utility.

pa() is left ifndef PRODUCT, as the class AllocatedObj is ifndef PRODUCT in src/hotspot/share/memory/allocation.hpp

pns(...) uses frame fr(sp, fp, pc); but we have an ifndef PRODUCT on frame constructor:  frame(void* sp, void* fp, void* pc);
Leave pns and pns2 ifndef PRODUCT.

File size changes:
Linux: libjvm.so becomes 8904 bytes larger.
Windows: jvm.dll becomes 6656 bytes larger.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8264593](https://bugs.openjdk.java.net/browse/JDK-8264593): debug.cpp utilities should be available in product builds.

### Reviewers
 * [Serguei Spitsyn](https://openjdk.java.net/census#sspitsyn) (@sspitsyn - **Reviewer**) ⚠️ Review applies to c3c81fc7c757c26b7c9112c4cee20cba04b3078e
 * [Coleen Phillimore](https://openjdk.java.net/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to c3c81fc7c757c26b7c9112c4cee20cba04b3078e
 * [Vladimir Ivanov](https://openjdk.java.net/census#vlivanov) (@iwanowww - **Reviewer**) ⚠️ Review applies to c3c81fc7c757c26b7c9112c4cee20cba04b3078e

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3307/head:pull/3307` \
`$ git checkout pull/3307`

Update a local copy of the PR: \
`$ git checkout pull/3307` \
`$ git pull https://git.openjdk.java.net/jdk pull/3307/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3307`

View PR using the GUI difftool: \
`$ git pr show -t 3307`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3307.diff">https://git.openjdk.java.net/jdk/pull/3307.diff</a>

</details>

### CODE REVIEW

**Review by sspitsyn — APPROVED** [review]
Hi Kevin,
Sorry for the latency.
This looks good to me. Thank you for taking care about it!
I vote for these changes as they improve debugability of product builds.
This also makes the debug.cpp functions potentially available for post-mortem analysis.
Thanks,
Serguei

File: src/hotspot/share/utilities/debug.cpp [hunk 1]
```diff
@@ -491,16 +489,16 @@ extern "C" void pp(void* p) {
...
 
 // pv: print vm-printable object
-extern "C" void pa(intptr_t p)   { ((AllocatedObj*) p)->print(); }
-extern "C" void findpc(intptr_t x);
-
-#endif // !PRODUCT
+#ifndef PRODUCT
+// AllocatedObj in allocation.hpp is not defined for PRODUCT
+extern "C" JNIEXPORT void pa(intptr_t p)   { ((AllocatedObj*) p)->print(); }
+#endif
```
Comment (coleenp) [inline src/hotspot/share/utilities/debug.cpp:495]: I don't really think 'pa' is interesting enough to keep.

**Review by coleenp — APPROVED** [review]
Looks good with a very minor comment.  Thank you for doing this and sorry for not getting to it sooner!

**Review by iwanowww — APPROVED** [review]
Looks good. It's a very valuable improvement in debuggability of product binaries. 

I'm surprised that all the diagnostic code debug commands rely on is already available in product binaries.

File: src/hotspot/share/utilities/debug.cpp (hunk 1 above)
Comment (kevinjwalls) [reply src/hotspot/share/utilities/debug.cpp:495]: Thanks Coleen - I haven't used pa either.  Will update without it...

Comment (kevinjwalls) [reply src/hotspot/share/utilities/debug.cpp]: I've taken out pa(intptr_t) and the misleading comment above it saying "pv".  
Also removed pv() from the help, which was perhaps an old name for pa.

**Comment by kevinjwalls** [comment]
Thanks all.  The main limit with product builds is that printing Java object details e.g. with hsfind() doesn't include full field information, though it's still useful.  Plan to consider that separately.

**Pull request closed without merging**